### PR TITLE
fix(rust, python): groupby rolling with negative offset

### LINKED ISSUE #9250: Wrong `groupby_rolling` result when `period=2d` and `offset=-3d`
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

When `period=2d` and `offset=-3d`, the `groupby_rolling` gives unexpected result. Possibly due to incorrect boundary handling.
Strangely, when offset is set to some other values like `offset=-4d` or `offset=-5d`, then results are correct.

### Reproducible example

```python
import polars as pl
from datetime import datetime
dcol = pl.date_range(datetime(2021,1,1), datetime(2021,1,4), '1d', eager=True)
vcol = np.arange(1, len(dcol)+1)
df = pl.DataFrame({'datetime': dcol, 'v': vcol})
df

```
┌─────────────────────┬─────┐
│ datetime            ┆ v   │
│ ---                 ┆ --- │
│ datetime[μs]        ┆ i64 │
╞═════════════════════╪═════╡
│ 2021-01-01 00:00:00 ┆ 1   │
│ 2021-01-02 00:00:00 ┆ 2   │
│ 2021-01-03 00:00:00 ┆ 3   │
│ 2021-01-04 00:00:00 ┆ 4   │
└─────────────────────┴─────┘
```
```py
df.groupby_rolling('datetime', period='2d', offset='-3d').agg(pl.col('v'))
```
Wrong output:
```
┌─────────────────────┬───────────┐
│ datetime            ┆ v         │
│ ---                 ┆ ---       │
│ datetime[μs]        ┆ list[i64] │
╞═════════════════════╪═══════════╡
│ 2021-01-01 00:00:00 ┆ [1]       │
│ 2021-01-02 00:00:00 ┆ [1, 2]    │
│ 2021-01-03 00:00:00 ┆ [1, 2, 3] │
│ 2021-01-04 00:00:00 ┆ [2, 3, 4] │
└─────────────────────┴───────────┘
```
```


### Expected behavior

```
┌─────────────────────┬───────────┐
│ datetime            ┆ v         │
│ ---                 ┆ ---       │
│ datetime[μs]        ┆ list[i64] │
╞═════════════════════╪═══════════╡
│ 2021-01-01 00:00:00 ┆ []        │
│ 2021-01-02 00:00:00 ┆ [1]       │
│ 2021-01-03 00:00:00 ┆ [1, 2]    │
│ 2021-01-04 00:00:00 ┆ [2, 3]    │
└─────────────────────┴───────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.0
Index type:  UInt32
Platform:    Linux-5.4.0-149-generic-x86_64-with-glibc2.31
Python:      3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0]

----Optional dependencies----
numpy:       1.24.3
pandas:      1.5.3
pyarrow:     12.0.0
connectorx:  <not installed>
deltalake:   <not installed>
fsspec:      2023.5.0
matplotlib:  3.6.3
xlsx2csv:    <not installed>
xlsxwriter:  <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #9250 

Adding
- a big parametric test, checking against a pure Python implementation (simple but slow)
- a unit test with hand-calculated (pen and paper) results

Perf implications:
- for `period` == `offset` (the default), nothing, this stays the same as before
- for `period > offset`, `groupby_values_iter_window_behind_t` is used instead of `groupby_values_iter_full_lookbehind`. This is documented to be slower, but at least it's correct. In any case, this strikes me as a rarer use case than the default

### CODE REVIEW

File: polars/polars-time/src/windows/groupby.rs [hunk 1]
```diff
@@ -465,7 +461,11 @@ pub(crate) fn groupby_values_iter<'a>(
     // we have a (partial) lookbehind window
     if offset.negative {
         // only lookbehind
-        if offset.nanoseconds() == period.nanoseconds() {
+        if (matches!(closed_window, ClosedWindow::Right | ClosedWindow::Both)
+            && offset.nanoseconds() > period.nanoseconds())
+            || (matches!(closed_window, ClosedWindow::Left | ClosedWindow::None)
+                && offset.nanoseconds() >= period.nanoseconds())
+        {
```
Comment (MarcoGorelli) [inline polars/polars-time/src/windows/groupby.rs:468]: for a full lookbehind, the window must be behind the given point

if `offset` == `period`, and `closed` includes the right endpoint, then the window would include the given point, and so it wouldn't be a full lookbehind but just a partial one. If `closed` if `'left'`, on the other hand, then it's always a full lookbehind, so long as `offset >= period`

File: polars/polars-time/src/windows/groupby.rs [hunk 2]
```diff
@@ -525,49 +525,47 @@ pub fn groupby_values(
 
     // we have a (partial) lookbehind window
     if offset.negative {
-        if offset.duration_ns() >= period.duration_ns() {
```
Comment (MarcoGorelli) [inline polars/polars-time/src/windows/groupby.rs:528]: here, previously there were two paths:

- `offset` >= `period`, `offset` < `period * 2`: groupby_values_iter_full_lookbehind
- `offset` >= `period`, `offset` >= `period * 2`: groupby_values_iter_window_behind_t
- `offset` < `period`: groupby_values_iter_partial_lookbehind

Just changing it to
- `offset` >= `period`: groupby_values_iter_window_behind_t
- `offset` < `period`: groupby_values_iter_partial_lookbehind

works, but the comments suggest `groupby_values_iter_window_behind_t` is less performant than `groupby_values_iter_full_lookbehind`. It is still possible to use the latter, but only in some cases - I need to look into this carefully

File: py-polars/tests/parametric/test_groupby_rolling.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from polars.testing.parametric.strategies import strategy_closed, strategy_time_unit
+from polars.utils.convert import _timedelta_to_pl_duration
+
+if TYPE_CHECKING:
+    from polars.type_aliases import ClosedInterval, TimeUnit
+
+
+@given(
+    period=st.timedeltas(min_value=timedelta(microseconds=0)),
+    offset=st.timedeltas(),
```
Comment (honno) [inline py-polars/tests/parametric/test_groupby_rolling.py:21]: FYI Hypothesis has a handy [`map()` method for strategies](https://hypothesis.readthedocs.io/en/latest/data.html#mapping). Current the type hints of `period`/`offset` are wrong in the signature anywho.

```suggestion
    offset=st.timedeltas().map(_timedelta_to_pl_duration)
```

(had this thought and was going to message you on slack, but thought I'd check open PRs first :sweat_smile:)

Comment (MarcoGorelli) [reply py-polars/tests/parametric/test_groupby_rolling.py]: nice, good one, thanks!

File: polars/polars-time/src/windows/groupby.rs (hunk 2 above)
Comment (ritchie46) [reply polars/polars-time/src/windows/groupby.rs:528]: I belive `groupby_values_iter_full_lookbehind` assumes that `t` is completely behind the window. So there are more cases where we have that besides `period == offset`. 

I will have to dive into it which cases it were again. Do you have on top of mind which predicate would inlcude all cases where `t` is full lookbehind?

This is beneficial as in that case we can parallelize over `t` and then look from that point backwards in the slice to find the window.

Comment (ritchie46) [reply polars/polars-time/src/windows/groupby.rs:517]: Right, let's first make it correct. We can try to find fast paths later if needed. :+1:

Thanks!

**Comment by MarcoGorelli** [comment]
thanks for your review, I'll think about how to improve performance in the offset > period case

and thanks Alex and Matt for reviewing + helping with hypothesis!

**Pull request merged**